sirepo/package_data/static/js/sirepo.js

```diff
@@ -1038,6 +1041,345 @@ SIREPO.app.factory('timeService', function() {
     return self;
 });
 
+SIREPO.app.service('sbatchLoginService', function($rootScope, appState, authState, errorService, requestSender, stringsService) {
+    const self = {};
+
+    // The implementation is a state maachine. Learn more about state machines:
+    // https://hackernoon.com/state-machines-can-help-you-solve-complex-programming-problems
+
+    // States and events are strings which are implemented as constants _e_.
+    // This is Hungarian Notation: https://en.wikipedia.org/wiki/Hungarian_notation
+    // To ensure all states and events are correctly spelled and accounted for.
+    // Clients of this service call functions which check the arguments are in _STATES or _EVENTS.
+
+    // For a diagram, see https://github.com/radiasoft/sirepo/pull/7055#issuecomment-2111318797
+    // This code will be slightly different, because it has been debugged.
+    // The _TRANSITIONS defines the state machine.
+
+
+    // Unlike classic state machines, events are $broadcast ("sbatchLoginEvent") and
+    // watchers call _EVENT_QUERIES to see if particular actions are required. This
+    // has two advantages. The dispatcher is angular's normal digest loop, which avoids
+    // callbacks within an angular event. Second, _EVENT_QUERIES collects all the
+    // different and overlapping actions in different parts of the code base.
+
+    // The actions for sending/receiving messages to the server are handled
+    // by this service, since that was the original point of this service. These
+    // actions could be moved out.
+
+    // _STATE_QUERIES are used by clients outside of a transition (sbatchLoginEvent)
+    // to decide how to render.
+
+    // sbatchLogin was sent to server
+    const _s_auth = 'auth';
```

Comment:
I tried a longer mode, but this fills the table with way too much repetitive text, obscuring the meaning. Given that this is internal to the service, I consider this fine, since it's quite localized.